File: src/lib/languages.js
// Presenting `scan_languages` results, and caching them.
//
// The Rust side counts; everything here is about saying it in the smallest
// space that is still honest. Two rules drive the shapes below, both from
// docs/ROADMAP.md's language section:
//
//   * The sidebar is a list, not a report. A project line gets a couple of
//     words, and the full breakdown lives in its tooltip — depth on demand,
//     no new permanent chrome.
//   * A partial answer says it is partial. `truncated` is not decoration.

// A second cache rather than a generalisation of status-cache.js: the two
// invalidate at different moments. A map status changes when `generate`
// runs; a language count changes when files are added to the tree, which
// this app never observes and cannot invalidate on. Merging them would mean
// one of the two gets the wrong lifetime.
const cache = new Map();

/** The in-flight or resolved `scan_languages` result for `root`, fetched once. */
export function scanLanguages(invoke, root, mapDir) {
  let entry = cache.get(root);
  if (!entry) {
    entry = invoke("scan_languages", { root, mapDir: mapDir ?? null }).catch((e) => {
      // Same reasoning as status-cache.js: a failed fetch must not become a
      // permanently cached rejection.
      cache.delete(root);
      throw e;
    });
    cache.set(root, entry);
  }
  return entry;
}

/** Call after anything that changes which files are under `root`. */
export function invalidateLanguages(root) {
  cache.delete(root);
}

/**
 * The languages worth naming, largest first, with their share of counted
 * source files.
 *
 * `minPercent` exists because a tree with one stray `.sh` in it is not "a
 * Shell project", and listing it alongside the real answer makes the real
 * answer harder to find. It is a display rule, not a claim that the file is
 * not there — `rest` reports what was folded away.
 */
export function topShares(scan, { max = 3, minPercent = 5 } = {}) {
  if (!scan || !scan.total) return { shares: [], rest: 0 };

  const shares = [];
  let rest = 0;
  for (const lang of scan.languages) {
    const percent = Math.round((lang.files / scan.total) * 100);
    if (shares.length < max && percent >= minPercent) {
      shares.push({ name: lang.name, files: lang.files, percent });
    } else {
      rest += lang.files;
    }
  }
  return { shares, rest };
}

/**
 * One short line for a sidebar entry: the two dominant languages, no
 * percentages.
 *
 * Two, not three: this shares a line with the module/file counts, and a
 * third name is where it starts wrapping. The count of what was left out is
 * kept rather than dropped, so the line never implies the list is complete.
 */
export function badgeText(scan) {
  const { shares } = topShares(scan, { max: 2, minPercent: 10 });
  if (!shares.length) return null;

  const named = shares.map((s) => s.name);
  const remaining = (scan.languages?.length ?? 0) - named.length;
  return remaining > 0 ? `${named.join(" · ")} +${remaining}` : named.join(" · ");
}

/**
 * The full breakdown, for a tooltip or the folder-picker preview.
 *
 * Returns a string rather than DOM so both callers can decide where to put
 * it, and so this stays testable without a document.
 *
 * `support` is optional: without it the line reports what is there, with it
 * the line also reports what the engine will do about it. Omitted rather
 * than required so the count is still shown while the engine probe is in
 * flight or has failed -- a partial answer beats no line at all.
 */
export function summaryText(scan, support) {
  if (!scan) return "";
  if (!scan.total) {
    return "No recognised source files — nothing here for the engine to map.";
  }

  const { shares, rest } = topShares(scan, { max: 6, minPercent: 1 });
  const parts = shares.map((s) => {
    // Only the two states that change what the reader should expect are
    // marked. Annotating "full" too would put a tag on every entry of a
    // normal project and turn the line into a legend -- the note exists for
    // the exception, not as a per-language verdict.
    const state = support?.get(s.name);
    const note =
      state === "none" ? " — no backend" : state === "degraded" ? " — no grammar" : "";
    return `${s.percent} % ${s.name} (${s.files})${note}`;
  });
  if (rest > 0) parts.push(`${rest} more`);

  let line = parts.join(" · ");
  if (scan.truncated) {
    // Said plainly rather than with an ellipsis: a reader who does not know
    // the walk stopped early has no way to tell this apart from a complete
    // answer, which is the one reading that would be wrong.
    line += " — counted from a partial walk of a very large tree";
  }
  return line;
}

/**
 * What the engine can do with each language the count found.
 *
 * Joined on the tree-sitter grammar name, which both sides know
 * independently — the engine reports it in `--capabilities`, the counter
 * carries it per language, and neither had to be told the other's
 * vocabulary. Joining on backend names instead would mean this file knowing
 * that the engine calls TypeScript `ts`, which is exactly the capability
 * duplication the Rust side refuses to commit.
 *
 * Four states, and the fourth is the one worth spelling out:
 *
 *   "full"      a backend, and its grammar loaded — function-level data
 *   "degraded"  a backend whose grammar is missing — module tree only
 *   "none"      no backend for this language at all
 *   "unknown"   this engine cannot be asked (predates the field)
 *
 * `unknown` is not `none`. An older engine that reads Lua perfectly well
 * would be reported as unable to read anything, and the advice that follows
 * from the two is opposite: one is "this will not work", the other is "this
 * works, it just cannot explain itself".
 */
export function supportFor(scan, engineLanguages) {
  const known = engineLanguages?.languages ?? null;
  const out = new Map();
  if (!scan) return out;

  for (const lang of scan.languages) {
    if (!known) {
      out.set(lang.name, "unknown");
      continue;
    }
    // A backend can be registered under several names for one grammar, and
    // a language can in principle map to none — `find` over a handful of
    // entries is cheaper than building an index for a list this short.
    //
    // The second key is for a backend that reads a language with no parser
    // at all — assembly, whose three syntaxes are a fork rather than
    // dialects, so the engine reads it by line. It has no grammar name to be
    // found by, and joining on the grammar alone reported "no backend" for a
    // language the engine reads at full fidelity. The counter names the
    // backend only in that case; see `LanguageCount::backend` on the Rust
    // side for why that stays an exception rather than a second vocabulary.
    const backend = lang.grammar
      ? known.find((b) => b.grammar === lang.grammar)
      : lang.backend
        ? known.find((b) => b.name === lang.backend)
        : undefined;
    if (!backend) {
      out.set(lang.name, "none");
    } else {
      // `grammar_loaded === false` is a real answer; `null`/absent means the
      // backend needs no parser, which is full fidelity, not a degradation.
      out.set(lang.name, backend.grammar_loaded === false ? "degraded" : "full");
    }
  }
  return out;
}

/**
 * Whether anything in this project has call extraction behind it.
 *
 * The Calls and Module Calls views render empty for nineteen of the
 * engine's twenty-three backends, and identically to how they render for a
 * project that genuinely has no calls in it. Those are different facts. This
 * is the join that lets the window say which — the same join `supportFor`
 * does for grammars, over the same two inputs.
 *
 * Three answers, and the third is not a rounding of the second:
 *
 * - `"yes"` — at least one language here produces call edges, so an empty
 *   panel is about the project.
 * - `"no"` — none of them do, so an empty panel is about the build.
 * - `"unknown"` — there is no scan yet, or the engine predates the field.
 *   Saying nothing is right there: a note that guessed would be wrong for
 *   exactly the reader who most needs it to be right.
 *
 * "At least one" rather than "all", deliberately. A repository with Lua
 * beside Go does have a call graph — a partial one — and telling its reader
 * the feature is unsupported would be false. The unhelpful case (a partial
 * graph that looks whole) is a different problem and not one a note over an
 * empty panel can fix, because that panel is not empty.
 *
 * @param {{languages: {name: string, grammar: string|null, backend?: string|null}[]}|null} scan
 * @param {{languages: {name: string, grammar: string|null, calls?: boolean|null}[]|null}|null} engineLanguages
 * @returns {"yes"|"no"|"unknown"}
 */
export function callsSupportFor(scan, engineLanguages) {
  const known = engineLanguages?.languages ?? null;
  if (!scan || !known || scan.languages.length === 0) return "unknown";

  let sawAnswer = false;
  for (const lang of scan.languages) {
    const backend = lang.grammar
      ? known.find((b) => b.grammar === lang.grammar)
      : lang.backend
        ? known.find((b) => b.name === lang.backend)
        : undefined;
    if (!backend) continue;
    // Absent on an engine that predates the field. Skipped rather than read
    // as `false`, so one old answer cannot turn the whole verdict into a
    // claim the engine never made.
    if (backend.calls === undefined || backend.calls === null) continue;
    sawAnswer = true;
    if (backend.calls === true) return "yes";
  }
  return sawAnswer ? "no" : "unknown";
}

/**
 * The engine's own backend list, for the Engine panel.
 *
 * Grouped rather than annotated per entry. Repeating "(no grammar — module
 * tree only)" after each of three backends says one thing three times and
 * pushes the line past the panel width; the caveat is a property of a set
 * here, not of each member.
 */
export function engineLanguageText(engineLanguages) {
  const known = engineLanguages?.languages ?? null;
  if (!known) {
    return "This engine is older than the language list — it cannot say which languages it reads.";
  }
  if (!known.length) return "This engine reports no language backends at all.";

  // `grammar_loaded === false` is the only degraded state: `null` means the
  // backend needs no parser, which is full fidelity.
  const full = known.filter((b) => b.grammar_loaded !== false).map((b) => b.name);
  const missing = known.filter((b) => b.grammar_loaded === false).map((b) => b.name);

  if (!missing.length) return full.join(" · ");
  const tail = `no grammar for ${missing.join(", ")} — module tree only`;
  return full.length ? `${full.join(" · ")} · ${tail}` : tail;
}

/**
 * The one-word verdict in the Engine panel's summary, which stays visible
 * when the panel is collapsed.
 *
 * Previously computed from whether a grammars *directory* resolved, which
 * is a proxy for the question and not the question. A directory holding
 * only `lua.dll` resolved, so the summary read "ready" while three of four
 * backends silently produced module-tree-only data — the exact silent
 * degradation this ecosystem treats as its most expensive failure mode,
 * sitting in the indicator meant to prevent it.
 *
 * Now: ask the engine, and fall back to the directory proxy only when the
 * engine is too old to be asked. The fallback is not an improvement on the
 * old behaviour and is not meant to be — it *is* the old behaviour, kept
 * for the case where nothing better is available, rather than reporting
 * "unknown" at a machine where the proxy was right all along.
 */
export function engineVerdict(engine, engineLanguages) {
  if (!engine?.path) return "not found";

  const known = engineLanguages?.languages ?? null;
  if (!known || !known.length) return engine.grammars ? "ready" : "no grammars";

  const wanting = known.filter((b) => b.grammar_loaded === false).length;
  if (!wanting) return "ready";
  if (wanting === known.length) return "no grammars";
  // Named as a count rather than "partial": the reader's next question is
  // "how much is missing", and a word that hides the number would only make
  // them open the panel to find out.
  return `${known.length - wanting} of ${known.length} grammars`;
}

/**
 * The diagnosis behind "no grammar for python, go, …".
 *
 * The verdict above answers *whether* fidelity is reduced. This answers the
 * question it provokes — **what do I put where** — which is the half that
 * was left when the grammar *manager* was rejected: downloading native
 * shared libraries from a rolling tag with no published checksum is a
 * silent update channel for unverified executable code, but saying which
 * file is missing from which directory costs nothing and reaches the
 * network not at all.
 *
 * Returns a catalog key and its parameters rather than a sentence, so the
 * only thing this function knows about language is that it does not speak
 * one. That is also `I18N.md`'s I18N-0 rule, applied early to the one place
 * here that was going to need it.
 *
 * **It does not repeat which backends are missing a grammar.** The line
 * directly above it in the panel already names them, in the engine's own
 * vocabulary; measured in a browser, saying it twice cost 141px of a 259px
 * sidebar column and said it in two vocabularies at once — backend names
 * there (`csharp`), grammar names here (`c_sharp`) — which reads as two
 * different problems rather than one fact.
 *
 * `null` means say nothing, and it covers two different silences that must
 * both stay silent: every backend has its grammar (nothing to diagnose),
 * and the engine is too old to be asked (`languages: null` — a guess here
 * would be wrong for exactly the reader who most needs it right).
 *
 * @param {{languages: ({name: string, grammar: string|null, grammar_loaded: boolean|null}[])|null}|null} engineLanguages
 * @param {{dir: string|null, from_setting: boolean, exists: boolean, files: string[], more: number}|null} dirInfo
 * @returns {{key: string, params: Record<string, string|number>}|null}
 */
export function grammarDiagnosis(engineLanguages, dirInfo) {
  const known = engineLanguages?.languages ?? null;
  if (!known || !known.length) return null;

  const missing = known.filter((b) => b.grammar_loaded === false);
  if (!missing.length) return null;

  // The grammar name, not the backend name: `csharp` is the backend and
  // `c_sharp` is the file. Naming the backend here would send somebody
  // looking for a file that is not what the engine asks for. Falls back to
  // the backend name only if an engine ever reports a wanted grammar
  // without naming it.
  const names = missing.map((b) => b.grammar || b.name);
  const files = dirInfo?.files || [];

  // **The example must be a file that is not already there.** Seen in a
  // browser before this rule existed: with nineteen backends wanting a
  // grammar and four files present, the sentence read "…named after the
  // grammar — `javascript.dll`" two clauses after listing `javascript.dll`
  // among what the directory holds. A backend can want a grammar it cannot
  // load for reasons that have nothing to do with the file being absent —
  // a bad build, the wrong architecture — so "wanted" and "missing from
  // this directory" are not the same set, and only the second one makes an
  // example worth copying.
  const absent = names.find((n) => !files.some((f) => f.startsWith(n + ".")));
  const params = { example: absent || names[0] };

  if (!dirInfo || !dirInfo.dir) return { key: "grammars.diag.none", params };
  if (!dirInfo.exists) return { key: "grammars.diag.gone", params: { ...params, dir: dirInfo.dir } };

  if (!files.length) {
    return { key: "grammars.diag.empty", params: { ...params, dir: dirInfo.dir } };
  }
  const have = dirInfo.more ? `${files.join(", ")} (+${dirInfo.more})` : files.join(", ");
  return { key: "grammars.diag.dir", params: { ...params, dir: dirInfo.dir, have } };
}

File: src-tauri/src/languages.rs
//! What languages a directory is written in, counted before anything is
//! generated.
//!
//! ## Why this exists at all
//!
//! The engine reads Lua and JavaScript/TypeScript today, with Python, Rust,
//! Go and C planned (see `documentation.nvim`'s
//! `docs/ROADMAP/IDEAS/MULTILANG.md`). Until every language a user points at
//! is supported, the interesting failure is not an error — it is a map that
//! comes back thin, with nothing on screen saying why. A repository that is
//! two thirds Python produces a perfectly valid, nearly empty map, and the
//! app currently reports that as success.
//!
//! Counting file extensions answers that before the first generate, at the
//! moment the folder is picked, when the answer is still cheap. It is
//! deliberately **not** a parse: no grammar, no engine, no subprocess. A
//! count of what is on disk is enough to say "this is mostly Python", and
//! anything more precise would be the engine's job, which this program does
//! not do.
//!
//! ## What this module does not decide
//!
//! **Whether a language can be read.** That is the engine's fact, not this
//! program's, and duplicating it here would be exactly the drift the whole
//! ecosystem exists to detect — an app claiming Python is unsupported after
//! the engine grew a Python backend is a worse failure than saying nothing.
//! So this returns names and counts only; the Engine panel's own
//! `--capabilities` answer supplies support, and the two are joined in the
//! frontend where both are already in hand.
//!
//! The extension-to-name table below is a *naming* fact ("`.rs` files are
//! called Rust"), which does not rot the way a capability list does.

use std::collections::HashMap;
use std::fs;
use std::path::Path;

use serde::Serialize;

/// One language found in a tree, with how many files carried it.
///
/// `grammar` is the join key against the engine's own `--capabilities`
/// answer, and it is a tree-sitter grammar name rather than an engine
/// backend name on purpose. The engine calls its TypeScript backend `ts`;
/// nothing outside the engine knows that, and hardcoding it here would be
/// the capability duplication this module's header refuses. A grammar name
/// is a third-party vocabulary both sides already speak independently --
/// the same class of fact as the extension-to-name table, and it rots the
/// same way that one does, which is to say not at all.
///
/// `None` for a language with no tree-sitter grammar worth naming here. The
/// frontend renders that as unknown rather than unsupported.
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct LanguageCount {
    pub name: String,
    pub files: u32,
    pub grammar: Option<String>,
}

/// The whole verdict for one directory.
///
/// `truncated` matters more than it looks: a count taken from a partial walk
/// is still useful ("mostly Python" does not become false at 20 000 files),
/// but presenting it as complete when it is not would be the silent
/// degradation this ecosystem treats as its most expensive failure mode. The
/// frontend says so when this is set.
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct LanguageScan {
    pub languages: Vec<LanguageCount>,
    /// Files that matched a known extension. The denominator for shares —
    /// **not** the number of files in the tree, which would make every share
    /// a fraction of README files and lockfiles too.
    pub total: u32,
    pub truncated: bool,
}

/// Stop before a big tree turns a folder-picker click into a visible pause.
///
/// 20 000 is chosen against the shape of the answer rather than a benchmark:
/// the question is "what is this repository mostly written in", and no
/// realistic tree changes its answer between the first 20 000 source files
/// and the last one. A cap that is generous enough to never truncate a real
/// project would be a cap that does not protect against the pathological
/// case it exists for (a home directory picked by accident).
const MAX_FILES: usize = 20_000;

/// Directories never descended into.
///
/// Two different reasons, deliberately in one list because the walk does not
/// care which applies:
///
/// * **Not the project's own code** — `node_modules`, `vendor`, `.venv`: a
///   tree with 40 000 dependency files is not "mostly TypeScript" because of
///   what its author wrote.
/// * **Generated output** — `target`, `dist`, `build`.
///
/// Two exclusions this list *cannot* express, both handled separately in
/// [`scan`]:
///
/// * **The map directory** is `docs/map`, a path rather than a name, and
///   skipping every directory called `map` would silently drop a real source
///   directory in some other project. Letting the act of generating a map
///   change the answer to "what is this project written in" would be absurd
///   on its face.
/// * **A checkout inside the checkout** has no fixed name at all — see
///   [`scan`]'s nested-repository rule, which was written after measuring
///   this walk against `documentation.nvim` and finding 306 of its 448 Lua
///   files were copies of itself under `.claude/worktrees/`.
const SKIP_DIRS: &[&str] = &[
    ".git",
    // Tooling state, never the project's own source. Listed by name as well
    // as caught by the nested-repository rule below, because the parts of it
    // that are *not* worktrees (settings, skills, agent definitions) are
    // still not what anyone means by "what is this project written in".
    ".claude",
    ".deps",
    ".direnv",
    "third_party",
    "bower_components",
    ".stack-work",
    ".dart_tool",
    ".hg",
    ".svn",
    ".jj",
    "node_modules",
    "vendor",
    "target",
    "dist",
    "build",
    "out",
    ".next",
    ".nuxt",
    ".svelte-kit",
    "__pycache__",
    ".venv",
    "venv",
    ".tox",
    ".mypy_cache",
    ".pytest_cache",
    ".ruff_cache",
    ".gradle",
    ".idea",
    ".vscode",
    ".cache",
    "coverage",
    ".terraform",
    "Pods",
    "DerivedData",
];

/// Extension (lowercase, no dot) to display name.
///
/// Broader than the engine can read on purpose — the whole point is to name
/// the languages it *cannot* read yet. Names are the conventional display
/// spelling ("C++", not "cpp"), because this string is shown to a person.
///
/// Related extensions collapse to one name where the distinction is not one a
/// reader is asking about (`.h` is C here; `.hpp` is C++), and stay apart
/// where it is (JavaScript vs. TypeScript, which are separate engine backends
/// and separate answers to "can this be mapped").
fn language_for(ext: &str) -> Option<(&'static str, Option<&'static str>)> {
    let (name, grammar) = match ext {
        "lua" => ("Lua", "lua"),
        "js" | "mjs" | "cjs" => ("JavaScript", "javascript"),
        "jsx" => ("JSX", "javascript"),
        "ts" | "mts" | "cts" => ("TypeScript", "typescript"),
        "tsx" => ("TSX", "tsx"),
        "py" | "pyi" => ("Python", "python"),
        "rs" => ("Rust", "rust"),
        "go" => ("Go", "go"),
        "c" | "h" => ("C", "c"),
        "cc" | "cpp" | "cxx" | "hpp" | "hh" | "hxx" => ("C++", "cpp"),
        "cs" => ("C#", "c_sharp"),
        "java" => ("Java", "java"),
        "kt" | "kts" => ("Kotlin", "kotlin"),
        "swift" => ("Swift", "swift"),
        "m" | "mm" => ("Objective-C", "objc"),
        "rb" => ("Ruby", "ruby"),
        "php" => ("PHP", "php"),
        "pl" | "pm" => ("Perl", "perl"),
        "sh" | "bash" | "zsh" => ("Shell", "bash"),
        "ps1" | "psm1" => ("PowerShell", "powershell"),
        "vim" => ("Vimscript", "vim"),
        "el" => ("Emacs Lisp", "elisp"),
        "ex" | "exs" => ("Elixir", "elixir"),
        "erl" | "hrl" => ("Erlang", "erlang"),
        "hs" => ("Haskell", "haskell"),
        "ml" | "mli" => ("OCaml", "ocaml"),
        "scala" | "sc" => ("Scala", "scala"),
        "clj" | "cljs" | "cljc" => ("Clojure", "clojure"),
        "dart" => ("Dart", "dart"),
        "zig" => ("Zig", "zig"),
        "nim" => ("Nim", "nim"),
        "jl" => ("Julia", "julia"),
        "r" => ("R", "r"),
        "sql" => ("SQL", "sql"),
        "vue" => ("Vue", "vue"),
        "svelte" => ("Svelte", "svelte"),
        "css" | "scss" | "sass" | "less" => ("CSS", "css"),
        "html" | "htm" => ("HTML", "html"),
        _ => return None,
    };
    Some((name, Some(grammar)))
}

/// Is this directory its own repository, rather than part of the one being
/// counted?
///
/// Found by measurement, not by reasoning: scanning `documentation.nvim`
/// reported 448 Lua files where `lua/` holds 98, and 306 of the difference
/// were full copies of the repository living under `.claude/worktrees/`.
/// A name list cannot catch that in general -- the next such directory will
/// be called something else -- but git already answers the question, and it
/// answers it for vendored submodule checkouts and stray clones too.
///
/// `.git` is tested for existence rather than for being a directory on
/// purpose: in a worktree and in a submodule it is a *file* pointing at the
/// real git directory elsewhere, which is precisely the case that matters
/// most here.
///
/// The root being scanned is never subjected to this -- the walk starts
/// inside it -- so a normal repository is not skipped by its own `.git`.
fn is_nested_checkout(dir: &Path) -> bool {
    dir.join(".git").exists()
}

/// Count source files by language under `root`.
///
/// `map_dir` is the project's generated-map directory, skipped wholesale for
/// the reason [`SKIP_DIRS`] explains. `None` falls back to the conventional
/// `docs/map` under `root` — the same default `add_project` writes into a new
/// `Project`, so a folder being previewed before it is added gets the same
/// treatment as one already in the list.
///
/// Iterative rather than recursive, and symlinks are never followed: a
/// symlink back up the tree is a cycle, and a cycle in a recursive walk is a
/// stack overflow rather than a slow answer.
pub fn scan(root: &Path, map_dir: Option<&Path>) -> Result<LanguageScan, String> {
    if !root.is_dir() {
        return Err(format!("{} is not a directory", root.display()));
    }

    // Compared as an owned path rather than by string: the caller's `map_dir`
    // arrives from the workspace with forward slashes even on Windows (see
    // `add_project`'s normalisation), and `PathBuf` comparison handles that
    // where a `==` on the two strings would not.
    let skip_map = map_dir
        .map(|p| p.to_path_buf())
        .unwrap_or_else(|| root.join("docs").join("map"));

    let mut counts: HashMap<&'static str, (u32, Option<&'static str>)> = HashMap::new();
    let mut visited = 0usize;
    let mut truncated = false;
    let mut stack = vec![root.to_path_buf()];

    while let Some(dir) = stack.pop() {
        // An unreadable subdirectory is not a reason to fail the whole scan.
        // A permission-denied node_modules two levels down should cost its own
        // contents, not the answer for everything above it.
        let entries = match fs::read_dir(&dir) {
            Ok(e) => e,
            Err(_) => continue,
        };

        for entry in entries.flatten() {
            if visited >= MAX_FILES {
                truncated = true;
                break;
            }

            // `file_type` rather than `metadata`: it does not follow symlinks,
            // which is what makes the cycle guarantee above true.
            let ft = match entry.file_type() {
                Ok(t) => t,
                Err(_) => continue,
            };
            if ft.is_symlink() {
                continue;
            }

            let name = entry.file_name().to_string_lossy().to_string();

            if ft.is_dir() {
                let path = entry.path();
                if !SKIP_DIRS.contains(&name.as_str())
                    && path != skip_map
                    && !is_nested_checkout(&path)
                {
                    stack.push(path);
                }
                continue;
            }

            visited += 1;
            let ext = match name.rsplit_once('.') {
                Some((_, ext)) if !ext.is_empty() => ext.to_lowercase(),
                _ => continue,
            };
            if let Some((lang, grammar)) = language_for(&ext) {
                let slot = counts.entry(lang).or_insert((0, grammar));
                slot.0 += 1;
            }
        }

        if truncated {
            break;
        }
    }

    let total: u32 = counts.values().map(|(n, _)| n).sum();
    let mut languages: Vec<LanguageCount> = counts
        .into_iter()
        .map(|(name, (files, grammar))| LanguageCount {
            name: name.to_string(),
            files,
            grammar: grammar.map(str::to_string),
        })
        .collect();

    // Descending by count, then by name — the tie-break is what keeps this
    // deterministic. A `HashMap` iterates in an arbitrary order that can
    // differ between runs of the same binary, and two languages with equal
    // counts swapping places between renders would look like a bug.
    languages.sort_by(|a, b| b.files.cmp(&a.files).then_with(|| a.name.cmp(&b.name)));

    Ok(LanguageScan {
        languages,
        total,
        truncated,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    /// One expected row. Spelling the grammar out at every call site rather
    /// than deriving it from `language_for` -- a test that computed the
    /// expected value the same way the code does would pass with the join
    /// key wrong, which is the one thing these assertions exist to catch.
    fn lc(name: &str, files: u32, grammar: &str) -> LanguageCount {
        LanguageCount {
            name: name.into(),
            files,
            grammar: Some(grammar.into()),
        }
    }

    /// A throwaway tree under the OS temp dir. Named per test so two tests
    /// running in parallel (which `cargo test` does by default) cannot see
    /// each other's files.
    fn tree(name: &str, files: &[&str]) -> std::path::PathBuf {
        let root = std::env::temp_dir().join(format!("docmap-langs-{name}"));
        let _ = fs::remove_dir_all(&root);
        for rel in files {
            let path = root.join(rel);
            fs::create_dir_all(path.parent().unwrap()).unwrap();
            fs::write(&path, "").unwrap();
        }
        fs::create_dir_all(&root).unwrap();
        root
    }

    #[test]
    fn counts_by_language_and_ranks_by_count() {
        let root = tree(
            "rank",
            &["a.py", "b.py", "c.py", "d.rs", "e.rs", "f.lua"],
        );
        let scan = scan(&root, None).unwrap();
        assert_eq!(
            scan.languages,
            vec![
                lc("Python", 3, "python"),
                lc("Rust", 2, "rust"),
                lc("Lua", 1, "lua"),
            ]
        );
        assert_eq!(scan.total, 6);
        assert!(!scan.truncated);
    }

    #[test]
    fn equal_counts_break_the_tie_by_name_so_the_order_is_stable() {
        let root = tree("tie", &["a.go", "b.rs"]);
        let first = scan(&root, None).unwrap();
        let second = scan(&root, None).unwrap();
        assert_eq!(first, second);
        assert_eq!(first.languages[0].name, "Go");
    }

    #[test]
    fn skips_dependency_and_generated_directories() {
        let root = tree(
            "skip",
            &[
                "src/main.rs",
                "node_modules/dep/index.js",
                "target/debug/thing.rs",
                "docs/map/index.html",
                ".git/hooks/pre-commit.sh",
            ],
        );
        let scan = scan(&root, None).unwrap();
        // Only `src/main.rs` survives: the three dependency/generated trees
        // are skipped by name, and `docs/map` by the separate path check --
        // its 1.5 MB index.html would otherwise make every mapped project
        // partly "HTML" the moment it was generated.
        assert_eq!(
            scan.languages,
            vec![lc("Rust", 1, "rust")]
        );
    }

    #[test]
    fn an_explicit_map_dir_is_skipped_instead_of_the_conventional_one() {
        let root = tree(
            "mapdir",
            &["a.lua", "docs/map/index.html", "elsewhere/index.html"],
        );
        // Told the map lives in `elsewhere`, the walk must skip that and stop
        // skipping `docs/map` -- otherwise a project configured away from the
        // convention silently gets both wrong.
        let scan = scan(&root, Some(&root.join("elsewhere"))).unwrap();
        assert_eq!(
            scan.languages,
            vec![
                lc("HTML", 1, "html"),
                lc("Lua", 1, "lua"),
            ]
        );
    }

    #[test]
    fn a_directory_that_is_its_own_checkout_is_not_counted() {
        let root = tree(
            "nested",
            &["a.lua", "worktrees/copy/b.lua", "worktrees/copy/c.lua"],
        );
        // A git worktree marks itself with a `.git` *file*, not a directory --
        // the case this rule exists for. Without it, a repository with two
        // worktrees under it reports three times its own size.
        fs::write(root.join("worktrees/copy/.git"), "gitdir: /elsewhere").unwrap();

        let scan = scan(&root, None).unwrap();
        assert_eq!(
            scan.languages,
            vec![lc("Lua", 1, "lua")]
        );
    }

    #[test]
    fn files_without_an_extension_or_with_an_unknown_one_are_not_counted() {
        let root = tree("unknown", &["Makefile", "LICENSE", "notes.txt", "a.lua"]);
        let scan = scan(&root, None).unwrap();
        assert_eq!(scan.total, 1);
        assert_eq!(scan.languages[0].name, "Lua");
    }

    #[test]
    fn extensions_are_matched_case_insensitively() {
        let root = tree("case", &["A.LUA", "b.Py"]);
        let scan = scan(&root, None).unwrap();
        assert_eq!(scan.total, 2);
    }

    #[test]
    fn a_path_that_is_not_a_directory_is_an_error_not_an_empty_result() {
        let root = tree("notdir", &["a.lua"]);
        assert!(scan(&root.join("a.lua"), None).is_err());
    }
}
